dbms/src/Flash/Mpp/MPPTaskManager.cpp

```diff
@@ -86,39 +88,54 @@ MPPGatherTaskSetPtr MPPQuery::addMPPGatherTaskSet(const MPPGatherId & gather_id)
 
 void MPPTaskMonitor::waitAllMPPTasksFinish(const std::unique_ptr<Context> & context)
 {
-    // The maximum seconds TiFlash will wait for all current MPP tasks to finish before shutting down
+    // The maximum seconds TiFlash will wait for all current MPP tasks or disagg snapshots to finish before shutting down
     static constexpr const char * GRACEFUL_WAIT_SHUTDOWN_TIMEOUT = "flash.graceful_wait_shutdown_timeout";
     // The default value of flash.graceful_wait_shutdown_timeout
     static constexpr UInt64 DEFAULT_GRACEFUL_WAIT_SHUTDOWN_TIMEOUT = 600;
     auto graceful_wait_shutdown_timeout
         = context->getUsersConfig()->getUInt64(GRACEFUL_WAIT_SHUTDOWN_TIMEOUT, DEFAULT_GRACEFUL_WAIT_SHUTDOWN_TIMEOUT);
-    LOG_INFO(log, "Start to wait all MPP tasks to finish, timeout={}s", graceful_wait_shutdown_timeout);
+
+    bool is_disagg_storage_mode = context->getSharedContextDisagg()->isDisaggregatedStorageMode();
+    const String & wait_target = is_disagg_storage_mode ? "disagg snapshots" : "MPP tasks";
+    LOG_INFO(log, "Start to wait all {} to finish, timeout={}s", wait_target, graceful_wait_shutdown_timeout);
+
     UInt64 graceful_wait_shutdown_timeout_ms = graceful_wait_shutdown_timeout * 1000;
     Stopwatch watch;
     // The first sleep before checking to reduce the chance of missing MPP tasks that are still in the process of being dispatched
-    std::this_thread::sleep_for(std::chrono::seconds(1));
+    std::this_thread::sleep_for(std::chrono::seconds(2));
     bool all_tasks_finished = false;
     while (true)
     {
         auto elapsed_ms = watch.elapsedMilliseconds();
-        if (!all_tasks_finished)
+        if (is_disagg_storage_mode)
         {
-            std::unique_lock lock(mu);
-            if (monitored_tasks.empty())
-                all_tasks_finished = true;
+            if (context->getSharedContextDisagg()->wn_snapshot_manager->getActiveSnapshotCount() == 0)
```

Comment:
```suggestion
            // For write node under disagg arch, should wait snapshot being released. Otherwise compute nodes
            // may meet error when calling `FetchDisaggPages` on write nodes.
            if (context->getSharedContextDisagg()->wn_snapshot_manager->getActiveSnapshotCount() == 0)
```